ethexe/consensus/src/validator/producer.rs

```diff
@@ -0,0 +1,435 @@
+            (!block_commitments.is_empty() || !code_commitments.is_empty()).then_some(
+                BatchCommitment {
+                    block_commitments,
+                    code_commitments,
+                },
+            ),
+        )
+    }
+
+    fn aggregate_block_commitments_for_block(
+        ctx: &ValidatorContext,
+        block_hash: H256,
+    ) -> Result<Vec<BlockCommitment>, AggregationError> {
+        let commitments_queue = ctx
+            .db
+            .block_commitment_queue(block_hash)
+            .ok_or_else(|| anyhow!("Block {block_hash} commitment queue is not in storage"))?;
+
+        let mut commitments = Vec::new();
+
+        let predecessor_block = block_hash;
+
+        for block in commitments_queue {
+            if !ctx.db.block_computed(block) {
+                // This can happen when validator syncs from p2p network and skips some old blocks.
+                return Err(AggregationError::SomeBlocksInQueueAreNotComputed(block));
+            }
+
+            let outcomes = ctx
+                .db
+                .block_outcome(block)
+                .ok_or_else(|| anyhow!("Cannot get from db outcome for computed block {block}"))?;
+
+            let previous_committed_block =
+                ctx.db.previous_not_empty_block(block).ok_or_else(|| {
+                    anyhow!(
+                        "Cannot get from db previous committed block for computed block {block}"
+                    )
+                })?;
+
+            let header = ctx
+                .db
+                .block_header(block)
+                .ok_or_else(|| anyhow!("Cannot get from db header for computed block {block}"))?;
+
+            commitments.push(BlockCommitment {
+                hash: block,
+                timestamp: header.timestamp,
+                previous_committed_block,
+                predecessor_block,
+                transitions: outcomes,
+            });
+        }
+
+        Ok(commitments)
+    }
+
+    fn aggregate_code_commitments_for_block(
+        ctx: &ValidatorContext,
+        block_hash: H256,
+    ) -> Result<Vec<CodeCommitment>, AggregationError> {
+        let codes_queue = ctx
+            .db
+            .block_codes_queue(block_hash)
+            .ok_or_else(|| anyhow!("Computed block {block_hash} codes queue is not in storage"))?;
+
+        let mut commitments = Vec::new();
+        codes_queue
+            .into_iter()
+            .filter_map(|id| ctx.db.code_valid(id).map(|valid| (id, valid)))
+            .try_for_each(|(id, valid)| -> Result<()> {
+                let CodeInfo { timestamp, .. } = ctx
+                    .db
+                    .code_blob_info(id)
+                    .ok_or_else(|| anyhow!("Validated code {id} blob info is not in storage"))?;
+
+                commitments.push(CodeCommitment {
+                    id,
+                    timestamp,
+                    valid,
+                });
+
+                Ok(())
+            })?;
```

Comment:
```suggestion
        let mut commitments = codes_queue
            .into_iter()
            .filter_map(|id| {
            	let valid = ctx.db.code_valid(id)?;
            	Some((id, valid))
            })
            .map(|(id, valid)| -> Result<()> {
                let CodeInfo { timestamp, .. } = ctx
                    .db
                    .code_blob_info(id)
                    .ok_or_else(|| anyhow!("Validated code {id} blob info is not in storage"))?;

                Ok(CodeCommitment {
                    id,
                    timestamp,
                    valid,
                })
            })
            .collect()?;
```

Reply:
Interesting idea but do you think it's possible to apply `?` to `Vec<Result<T>>` ?

Reply:
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect

> collect() can also create instances of types that are not typical collections. For example, a [String](https://doc.rust-lang.org/std/string/struct.String.html) can be built from [char](https://doc.rust-lang.org/std/primitive.char.html)s, and an iterator of [Result<T, E>](https://doc.rust-lang.org/std/result/enum.Result.html) items can be collected into Result<Collection<T>, E>. See the examples below for more.